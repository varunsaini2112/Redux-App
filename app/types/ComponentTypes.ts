import { RootStackParamList } from ".";

type LinkType = {
	value: string;
	href: keyof RootStackParamList;
};

type LinkProps = LinkType & {
	onPress: (screenName: keyof RootStackParamList) => void;
};

export type { LinkType, LinkProps };
