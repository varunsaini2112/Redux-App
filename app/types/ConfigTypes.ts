type RootStackParamList = {
	Home: undefined;
	SignUp: undefined;
	Count: undefined;
	Login: undefined;
	ToDo: undefined;
};

export type { RootStackParamList };
