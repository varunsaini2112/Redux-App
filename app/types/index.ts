export * from "./ConfigTypes";

export * from "./ComponentTypes";
