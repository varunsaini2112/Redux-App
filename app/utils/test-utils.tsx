import React, { PropsWithChildren, ReactElement } from "react";
import { Provider } from "react-redux";
import { PreloadedState } from "@reduxjs/toolkit";
import { RenderOptions, render } from "@testing-library/react-native";

import { AppStore, RootState, setupStore } from "../redux/store";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}

export function renderWithProviders(
	ui: ReactElement,
	{
		preloadedState = {},
		// Automatically create a store instance if no store was passed in
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) {
	function Wrapper({ children }: PropsWithChildren<{}>): ReactElement {
		return <Provider store={store}>{children}</Provider>;
	}
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
