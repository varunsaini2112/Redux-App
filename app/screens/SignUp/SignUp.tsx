import { Alert, Pressable, Text, View } from "react-native";
import { SubmitHandler, useForm } from "react-hook-form";
import { StackScreenProps } from "@react-navigation/stack";

import { RootStackParamList } from "../../types";
import styles from "./SignUp.styles";
import { Button, H1, TextInputField } from "../../components";
import { baseStyles } from "../../config/appStyles";
import { VALIDATIONS } from "../../constants";

type FormValues = {
	fullName: string;
	email: string;
	password: string;
};

const SignUp: React.FC<StackScreenProps<RootStackParamList, "SignUp">> = ({
	navigation
}) => {
	const { handleSubmit, control } = useForm<FormValues>();

	const onPressSubmit: SubmitHandler<FormValues> = (data) => {
		Alert.alert(JSON.stringify(data));
	};

	return (
		<View style={styles.mainContainer}>
			<View style={styles.headingContainer}>
				<H1>SignUp Here</H1>
			</View>
			<View>
				<TextInputField
					name="fullName"
					label="Full Name"
					placeholder="Enter your full name"
					control={control}
					rules={VALIDATIONS.fullName}
				/>
				<TextInputField
					name="email"
					label="Email"
					placeholder="someone@here.xyz"
					control={control}
					rules={VALIDATIONS.email}
				/>
				<TextInputField
					name="password"
					label="Password"
					placeholder="Password here..."
					control={control}
					rules={VALIDATIONS.password}
					secureTextEntry={true}
				/>
				<Button onPress={handleSubmit(onPressSubmit)}>
					<Text style={baseStyles.buttonText}>Submit</Text>
				</Button>
			</View>
			<Pressable onPress={navigation.goBack} style={styles.centerTextContainer}>
				<Text style={styles.centerText}>Already have an account?</Text>
			</Pressable>
		</View>
	);
};

export default SignUp;
