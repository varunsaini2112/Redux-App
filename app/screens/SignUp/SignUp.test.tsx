import { RouteProp } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { act, fireEvent, screen } from "@testing-library/react-native";

import { RootStackParamList } from "../../types";
import { renderWithProviders } from "../../utils/test-utils";
import SignUp from "./SignUp";
import { Alert } from "react-native";

const navigation = {
	navigate: () => {},
	goBack: () => {}
} as StackNavigationProp<RootStackParamList, "SignUp">;
const route = {} as RouteProp<RootStackParamList, "SignUp">;

describe("SignUp", () => {
	it("should match snapshots", () => {
		const snapshot = renderWithProviders(
			<SignUp navigation={navigation} route={route} />
		);

		expect(snapshot).toMatchSnapshot();
	});

	it("should render SignUp Heading", () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		expect(screen.getByText("SignUp Here")).toBeOnTheScreen();
	});

	it("should show field errors on submit button", async () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		await act(async () => {
			fireEvent.press(screen.getByText("Submit"));
		});

		expect(screen.getByText("Full Name is required")).toBeOnTheScreen();
		expect(screen.getByText("Email is required")).toBeOnTheScreen();
		expect(screen.getByText("Password is required")).toBeOnTheScreen();
	});

	it("should show short name length error on invalid entry", async () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		fireEvent.changeText(
			screen.getByPlaceholderText("Enter your full name"),
			"T"
		);
		await act(async () => {
			fireEvent.press(screen.getByText("Submit"));
		});

		expect(screen.getByText("Name is too short")).toBeOnTheScreen();
	});

	it("should show long name length error on invalid entry", async () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		fireEvent.changeText(
			screen.getByPlaceholderText("Enter your full name"),
			"T".repeat(41)
		);
		await act(async () => {
			fireEvent.press(screen.getByText("Submit"));
		});

		expect(screen.getByText("Name is too long")).toBeOnTheScreen();
	});

	it("should show name pattern error on invalid entry", async () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		fireEvent.changeText(
			screen.getByPlaceholderText("Enter your full name"),
			"Test@Name"
		);
		await act(async () => {
			fireEvent.press(screen.getByText("Submit"));
		});

		expect(screen.getByText("Please enter a valid name")).toBeOnTheScreen();
	});

	it("should submit successfully and show alert on valid field entry", async () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		const alertSpy = jest.spyOn(Alert, "alert");

		fireEvent.changeText(
			screen.getByPlaceholderText("Enter your full name"),
			"Test Name"
		);
		fireEvent.changeText(
			screen.getByPlaceholderText("someone@here.xyz"),
			"testField@example.com"
		);
		fireEvent.changeText(
			screen.getByPlaceholderText("Password here..."),
			"Test@12345"
		);
		await act(async () => {
			fireEvent.press(screen.getByText("Submit"));
		});

		screen.debug();
		expect(alertSpy).toHaveBeenCalled();
	});

	it("should render already have account text", () => {
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		expect(screen.getByText("Already have an account?")).toBeOnTheScreen();
	});

	it("should navigate to signup on press already have account text", () => {
		const navigationSpy = jest.spyOn(navigation, "goBack");
		renderWithProviders(<SignUp navigation={navigation} route={route} />);

		fireEvent.press(screen.getByText("Already have an account?"));

		expect(navigationSpy).toHaveBeenCalled();
	});
});
