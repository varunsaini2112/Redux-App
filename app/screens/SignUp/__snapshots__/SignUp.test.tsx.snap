// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should match snapshots 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "marginVertical": 24,
        "paddingVertical": 24,
      }
    }
  >
    <Text
      style={
        {
          "color": "#132631",
          "fontSize": 24,
          "fontWeight": "600",
        }
      }
    >
      SignUp Here
    </Text>
  </View>
  <View>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#132631",
              "fontSize": 14,
            }
          }
        >
          Full Name
        </Text>
      </View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Enter your full name"
        style={
          {
            "borderRadius": 16,
            "borderWidth": 0.4,
            "padding": 16,
          }
        }
        value=""
      />
      <View
        style={
          {
            "alignItems": "flex-end",
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontSize": 12,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#132631",
              "fontSize": 14,
            }
          }
        >
          Email
        </Text>
      </View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="someone@here.xyz"
        style={
          {
            "borderRadius": 16,
            "borderWidth": 0.4,
            "padding": 16,
          }
        }
        value=""
      />
      <View
        style={
          {
            "alignItems": "flex-end",
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontSize": 12,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "marginBottom": 16,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#132631",
              "fontSize": 14,
            }
          }
        >
          Password
        </Text>
      </View>
      <TextInput
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Password here..."
        secureTextEntry={true}
        style={
          {
            "borderRadius": 16,
            "borderWidth": 0.4,
            "padding": 16,
          }
        }
        value=""
      />
      <View
        style={
          {
            "alignItems": "flex-end",
          }
        }
      >
        <Text
          style={
            {
              "color": "red",
              "fontSize": 12,
            }
          }
        />
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "opacity": 1,
          },
          {
            "alignItems": "center",
            "backgroundColor": "#D80064",
            "borderRadius": 16,
            "marginVertical": 8,
            "padding": 16,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 20,
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "marginVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "textDecorationLine": "underline",
        }
      }
    >
      Already have an account?
    </Text>
  </View>
</View>
`;
